core/endorser/endorser.go

```diff
@@ -516,42 +535,105 @@ func (e *Endorser) buildChaincodeInterest(simResult *ledger.TxSimulationResults)
+	for _, collections := range chaincodes {
+		if len(collections) > 1 && collections[""] {
+			delete(collections, "")
 		}
 	}
 
 	// now build the chaincode interest
 	ccInterest := &pb.ChaincodeInterest{}
 	for chaincode, collections := range chaincodes {
-		if collections == nil {
-			ccInterest.Chaincodes = append(ccInterest.Chaincodes, &pb.ChaincodeCall{
+		keyPolicies := policies.keySignaturePolicies(chaincode)
+		sbeAttached := false
+		for collectionName, isRead := range collections {
+			ccCall := &pb.ChaincodeCall{
 				Name: chaincode,
-			})
-		} else {
-			for collectionName, isRead := range collections {
+			}
+			if collectionName != "" { // the empty collection name here represents the public RWset
 				// Since each collection in a chaincode could have different values of the NoPrivateReads flag, create a new Chaincode entry for each.
-				ccInterest.Chaincodes = append(ccInterest.Chaincodes, &pb.ChaincodeCall{
-					Name:            chaincode,
-					CollectionNames: []string{collectionName},
-					NoPrivateReads:  !isRead,
-				})
+				ccCall.CollectionNames = []string{collectionName}
+				ccCall.NoPrivateReads = !isRead
+			}
+			if !sbeAttached && len(keyPolicies) > 0 {
+				ccCall.KeyPolicies = keyPolicies
+				sbeAttached = true
+				// if there are no non-SBE public writes for this chaincode, then disregard the chaincode policy
+				if collectionName == "" && !policies.requireChaincodePolicy(chaincode) {
+					ccCall.DisregardNamespacePolicy = true
+				}
 			}
+			ccInterest.Chaincodes = append(ccInterest.Chaincodes, ccCall)
 		}
 	}
 
-	// add the key signature policies - at the moment they are grouped together in the first chaincode
-	// TODO implement DisregardNamespacePolicy hint in ChaincodeCall
-	if len(ccInterest.Chaincodes) > 0 {
-		for _, policyBytes := range simResult.KeySignaturePolicies {
-			policy, err := protoutil.UnmarshalSignaturePolicy(policyBytes)
-			if err != nil {
-				return nil, err
+	endorserLogger.Info("ccInterest", ccInterest)
+	return ccInterest, nil
+}
+
+type metadataPolicies struct {
+	sbePolicies    map[string][]*common.SignaturePolicyEnvelope
+	policyRequired map[string]map[string]bool
+}
+
+func parseWritesetMetadata(metadata ledger.WritesetMetadata) (*metadataPolicies, error) {
+	mp := &metadataPolicies{
+		sbePolicies:    map[string][]*common.SignaturePolicyEnvelope{},
+		policyRequired: map[string]map[string]bool{},
+	}
+	for ns, cmap := range metadata {
+		mp.policyRequired[ns] = map[string]bool{}
+		sbePolicies := []*common.SignaturePolicyEnvelope{}
+		for coll, kmap := range cmap {
+			// look through each of the states that were written to
+			for _, md := range kmap {
+				// if the metadata is nil, or it doesn't contain data relating to SBE policy, then this chaincode/collection policy is required
+				if md == nil {
+					mp.policyRequired[ns][coll] = true
+				}
```

Comment:
Nit: No functional difference but adds clarity.
```suggestion
				if md == nil {
					mp.policyRequired[ns][coll] = true
					continue
				}
```